pkg/apis/autoscaling/validation/validation.go

```diff
@@ -442,6 +442,14 @@ func validateMetricTarget(mt autoscaling.MetricTarget, fldPath *field.Path) fiel
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("type"), mt.Type, "must be either Utilization, Value, or AverageValue"))
 	}
 
+	if mt.Type == autoscaling.ValueMetricType && mt.Value == nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("value"), mt.Value, "must be set for metric target type Value"))
+	}
+
+	if mt.Type == autoscaling.AverageValueMetricType && mt.AverageValue == nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("averageValue"), mt.AverageValue, "must be set for metric target type AverageValue"))
+	}
```

Comment:
The new validation needs to be conditional and only run when creating new objects or updating objects that already pass this validation.

The pattern for doing that is to define something like:
```go
type HPAValidationOptions struct {
  AllowMismatchedTargetTypeAndValue bool
}
```
pass it to ValidateHorizontalPodAutoscaler and ValidateHorizontalPodAutoscalerUpdate, and plumb it all the way down to this function.

When calling from autoscalerStrategy#Validate, set AllowMismatchedTargetTypeAndValue to false.

When calling from autoscalerStrategy#ValidateUpdate, set AllowMismatchedTargetTypeAndValue to true IFF the old object has mismatched target type/value fields. This allows tolerating updates of existing persisted data that is already bad.

Reply:
I think there are three cases we need to check, both positive (associated value field is not nil) and negative (other value fields are nil):

```go
if !opts.AllowMismatchedTargetTypeAndValue {
  switch mt.Type {
  case autoscaling.UtilizationMetricType:
    if mt.AverageUtilization == nil {
      ... required error
    }
    if mt.Value != nil {
      ... invalid error
    }
    if mt.AverageValue != nil {
      ... invalid error
    }

  case autoscaling.ValueMetricType:
    if mt.Value == nil {
      ... required error
    }
    if mt.AverageUtilization != nil {
      ... invalid error
    }
    if mt.AverageValue != nil {
      ... invalid error
    }

  case autoscaling.AverageValueMetricType:
    if mt.AverageValue == nil {
      ... required error
    }
    if mt.Value != nil {
      ... invalid error
    }
    if mt.AverageUtilization != nil {
      ... invalid error
    }
  }
}
```

Reply:
Thanks for the feedback Jordan, I really appreciate it. I took a closer look at how the metric types tie into the metric sources and values being set. It looks like only `ObjectMetricSource` checks the target types before using the corresponding value, and I didn't really see target types being referenced in other places. The others don't seem to do this and there's a mix of usage which made it hard to generate a generic validator for all cases. So I incorporated the above changers only when checking `ObjectMetricSource`. 
For reference:
computeStatusForObjectMetric can use either Value or AverageValue
computeStatusForPodsMetric only uses AverageValue
computeStatusForResourceMetric and computeStatusForContainerResourceMetric can use either AverageValue or AverageUtilization
computeStatusForExternalMetric can use either Value or AverageValue